design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,228 @@
+the format for these profiles is vendor-specific eliminates a lot of the
+implementation complexity, at the risk of moving that complexity to the
+end-user. This requirement is similar to something Dell has proposed in the
+Ironic community
+([hw config](https://review.opendev.org/c/openstack/ironic-specs/+/740721))
+to export and import an entire system iDRAC inventory including RAID and
+networking requirements. Compared to that approach, this proposal attempts
+a more lightweight, vendor-neutral approach just for BIOS configuration.
+
+### Goals
+
+Provide a vendor-neutral approach of updating BIOS configuration in a bulk
+manner across all similar hardware. Create a template for the current
+BIOS configurations settings cached on a server. Give access to the BIOS
+registry which can be used as documentation so a user can easily update
+the BIOS settings to desired values. Validate the BIOS Configuration
+values based on the vendor's BIOS Registry before sending to the BMC.
+
+### Non-Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow the transfer of BIOS configuration between different vendor’s
+  servers or different vendor models
+- Support bulk set of other configurations besides BIOS
+- Provide storage mechanism for BIOS settings
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When provisioning a large amount of machines with identical settings, our
+customers want to provide a known, validated payload with all the desired
+firmware configuration.
+
+This ensures all the machines will have the identical settings, reducing issues
+caused by unexpected configuration differences.
+
+Since the name/value pair mapping is vendor specific and not intuitive, it
+doesn’t lend itself to creating the input by hand using existing vendor
+documentation, except for potentially a small subset of names. It’s desirable
+to take an existing BIOS config set and modify the current settings to the
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes a new Custom Resource Definition (CRD) to store the settings
+read from Ironic along with the modified settings to write to Ironic. Since
+each vendor provides 100 or more BIOS Settings, its not practical to store
+these in the BareMetalHost CRD. The new CRD will be named
+`HostFirmwareSettings` and will consist of the following:
+
+- `currentFirmwareSettings` - the current BIOS Settings retrieved from Ironic
+  via the bios API will be stored in the Status section. This dataset includes
+  BIOS Registry information which provides a schema for the settings.
+- `writeableFirmwareSettings` - BIOS Settings to write to the host via Ironic
+  will be stored in the Spec section. If it is empty, it will be initially
+  populated with the name/value pairs from `currentFirmwareSettings` that do
+  not have the `read-only` or `unique` flag set.
```

Comment:
I think we're ready for the bikeshed-on-naming portion of the discussion :grin:
Some thoughts:

* "desired" might be better than "writeable", but...
* it's kind of implied by the Spec/Status split that the Spec is the desired state and the Status is the current state
* the `status.currentFirmwareSettings` contains not only the settings but the whole schema. Calling them both "settings" might be confusing as it  implies they have the same structure.

What would folks think about something like `spec.vendorSettings` and `status.firmwareRegistry`?

Reply:
Thanks Zane! Agree that we can take "Firmware" out of the name since its redundant with that name in CRD itself, and also that using "settings" in both places is confusing.

For the spec section I think "vendorSettings" is good.
For the status section, this includes both the name+value and the registry section so I don't think calling it firmwareRegistry is adequate.  Maybe "currentValues"?

Reply:
Given that the name of the CRD is pretty specific and clear `HostFirmwareSettings`, what about reducing even further to `spec.vendorSettings` and `status.schema`?

@bfournie I think that in any case putting an example of the CRD help would be both useful as for documentation and to visualize it better for the ongoing discussion. For example, will the CRD have an `id` field to identify it? Do you plan to have additional fields for host selection?

Reply:
At any given time, how many CRD instances are expected to be present in the cluster (e.g. one for BMH, one for distinct vendor/server type, etc)? 
Also, considering that BMHs could be deployed in multiple namespaces, what would be the `scope` of the CRD?

Reply:
This resource is 1:1 with BMHs. They share the same name + namespace.

Reply:
I agree with y'all that the status one is tricky - we need to somehow convey that it contains both the current settings and the schema, ideally without literally typing currentSettingsAndSchema. In ironic 'registry' is taken to mean the schema, although in other contexts (e.g. Windows registry) it's the opposite, so it's confusing (or cunningly ambiguous, depending on your point of view).'values' has the opposite problem where it implies it's just the values.

I'm trying to think of words that describe values along with their schema, and the only thing coming to mind is 'database'. Maybe we should go with something generic like `settingsData` (or s/settings/configuration/ in both).

However, I also wonder if the fact that we are having difficulty coming up with a good name is a hint that we are trying too hard to squeeze two different things into one field? Given that we are in a separate CRD here I don't feel like there is the same premium on adding fields as there is in the BMH. Maybe we could afford to have separate `status.values` and `status.schema` fields? (A downside of this is that it makes it less clear that you should copy the `spec.vendorSettings`, not the `status.values`, when you want to clone settings to another Host.)

Reply:
I had thought about having a separate struct in the status.currentValues for the schema section, I do think it needs to be broken out as this schema doesn't (or shouldn't) change while the values will change when modified. The downside of having status.values and status.schema in separate fields though is we need the name field in both, if have something like this then you don't need to duplicate the name, but that doesn't really solve the naming problem:
currentValues {
string name
string value
struct schema {
string attributeType
etc.
}
}

Reply:
Names of spec and status fields have been updated and the CRD section has been included, let me know if this is reasonable.